<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" href="/css/main.css">
    <title>Sam Buckberry</title>
</head>
<body>
    <main>
        <header class="site-header">
            <h1 class="site-title"><a href="/">Sam Buckberry</a></h1>
            <nav class="site-nav">
                <a href="/about/" class="">About</a>
                <a href="/publications/" class="">Publications</a>
                <a href="/recipes/" class="">[R]ecipes</a>

            </nav>
        </header>
        <article class="post">
    <h2 class="post-header">RNA-Seq differential expression - FASTQ files to results using 10 lines of code</h2>
    <p class="post-meta">2015-07-14</p>
    <p>RNA sequencing, or <a href="https://en.wikipedia.org/wiki/RNA-Seq">RNA-Seq</a>, has rapidly become the preferred method of global differential gene expression analysis, yet many biologists are unfamiliar with how to analyse RNA-Seq data.</p>

<p>I&rsquo;m often asked by people starting out with RNA-Seq how they could begin to analyse their data. It&rsquo;s a tricky question because the methods vary substantially and the choices will depend on the experimental design, sequencing methodology and biological question(s) of interest.</p>

<p>However, there are several user-friendly R packages enable the those new to RNA-Seq analysis to easily get started with gene-level differential expression.</p>

<p>So to kick off, I&rsquo;ll show how you can go from raw data to results with 10 lines of code, then provide a brief breakdown the analysis step-by-step.</p>

<h2 id="fastq-files-to-differential-expression-results-using-10-lines-of-r-code">FASTQ files to differential expression results using 10 lines of R code</h2>
<div class="highlight"><pre><code class="language-r" data-lang="r">buildindex<span class="p">(</span>basename<span class="o">=</span><span class="s">&quot;chr1&quot;</span><span class="p">,</span> reference<span class="o">=</span><span class="s">&quot;chr1.fa.gz&quot;</span><span class="p">)</span>
align<span class="p">(</span>index<span class="o">=</span><span class="s">&quot;chr1&quot;</span><span class="p">,</span> readfile1<span class="o">=</span><span class="kp">list.files</span><span class="p">(</span>pattern<span class="o">=</span><span class="s">&quot;.fastq.gz$&quot;</span><span class="p">))</span>
fCounts <span class="o">&lt;-</span> featureCounts<span class="p">(</span>files<span class="o">=</span><span class="kp">list.files</span><span class="p">(</span>pattern<span class="o">=</span><span class="s">&quot;.BAM$&quot;</span><span class="p">),</span> annot.inbuilt<span class="o">=</span><span class="s">&quot;hg19&quot;</span><span class="p">)</span>
dge <span class="o">&lt;-</span> DGEList<span class="p">(</span>counts<span class="o">=</span>fCounts<span class="o">$</span>counts<span class="p">,</span> genes<span class="o">=</span>fCounts<span class="o">$</span>annotation<span class="p">)</span>
thresholdCPM <span class="o">&lt;-</span> <span class="kp">rowSums</span><span class="p">(</span>cpm<span class="p">(</span>dge<span class="p">)</span> <span class="o">&gt;</span> <span class="m">1</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="m">2</span>
dge <span class="o">&lt;-</span> dge<span class="p">[</span>thresholdCPM<span class="p">,</span> <span class="p">]</span>
design <span class="o">&lt;-</span> model.matrix<span class="p">(</span><span class="o">~</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;A&quot;</span><span class="p">,</span> <span class="s">&quot;A&quot;</span><span class="p">,</span> <span class="s">&quot;B&quot;</span><span class="p">,</span> <span class="s">&quot;B&quot;</span><span class="p">))</span>
voomDat <span class="o">&lt;-</span> voom<span class="p">(</span>dge<span class="p">,</span> design<span class="p">)</span>
fit <span class="o">&lt;-</span> eBayes<span class="p">(</span>lmFit<span class="p">(</span>voomDat<span class="p">,</span> design<span class="p">))</span>
topTable<span class="p">(</span>fit<span class="p">,</span> coef<span class="o">=</span><span class="m">2</span><span class="p">)</span>
</code></pre></div>
<h2 id="preliminaries">Preliminaries</h2>

<p><strong>Install R</strong> If you don&rsquo;t have R installed already, check out the <a href="http://www.r-project.org">R project</a>.</p>

<p><strong>Downlaod and install the required bioconductor packages</strong></p>
<div class="highlight"><pre><code class="language-r" data-lang="r"><span class="kn">source</span><span class="p">(</span><span class="s">&quot;http://bioconductor.org/biocLite.R&quot;</span><span class="p">)</span>
biocLite<span class="p">(</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;Rsubread&quot;</span><span class="p">,</span> <span class="s">&quot;limma&quot;</span><span class="p">,</span> <span class="s">&quot;edgeR&quot;</span><span class="p">))</span>
</code></pre></div>
<p><strong>Get some FASTQ files</strong>
If you need some data in <a href="https://en.wikipedia.org/wiki/FASTQ_format">FASTQ format</a> for testing purposes, you can download four files with 100,000 single-end reads hosted in <a href="https://github.com/SamBuckberry/Blog/tree/master/testFiles">this github repo</a> using the code below (~30MB).</p>
<div class="highlight"><pre><code class="language-r" data-lang="r">download.file<span class="p">(</span><span class="s">&quot;https://cdn.rawgit.com/SamBuckberry/Blog/master/testFiles/testFiles.zip&quot;</span><span class="p">,</span> destfile<span class="o">=</span><span class="s">&quot;testFiles.zip&quot;</span><span class="p">,</span> method<span class="o">=</span><span class="s">&quot;curl&quot;</span><span class="p">)</span>
unzip<span class="p">(</span><span class="s">&quot;testFiles.zip&quot;</span><span class="p">)</span>
</code></pre></div>
<p>Make sure that the FASTQ files are in your working directory.</p>

<p>In this example, we will only align reads to human chromosome 1 (chr1), but aligning to the whole genome would be no different.</p>

<p><strong>Download a reference genome in FASTA format</strong> </p>

<p>To download the chromosome 1 reference sequence in FASTA format from UCSC:</p>
<div class="highlight"><pre><code class="language-r" data-lang="r">download.file<span class="p">(</span><span class="s">&quot;http://hgdownload.soe.ucsc.edu/goldenPath/hg19/chromosomes/chr1.fa.gz&quot;</span><span class="p">,</span> destfile<span class="o">=</span><span class="s">&quot;chr1.fa.gz&quot;</span><span class="p">)</span>
untar<span class="p">(</span><span class="s">&quot;chr1.fa.gz&quot;</span><span class="p">)</span>
</code></pre></div>
<h2 id="step-by-step-breakdown-of-the-differential-expression-analysis">Step-by-step breakdown of the differential expression analysis</h2>

<p>First, load the required libraries:</p>
<div class="highlight"><pre><code class="language-r" data-lang="r"><span class="kn">library</span><span class="p">(</span>Rsubread<span class="p">)</span>
<span class="kn">library</span><span class="p">(</span>edgeR<span class="p">)</span>
<span class="kn">library</span><span class="p">(</span>limma<span class="p">)</span>
</code></pre></div>
<h3 id="step-1-build-a-reference-genome-index">Step 1: Build a reference genome index</h3>

<p>First, we need to create an index for the reference genome (This might take a couple of minutes).</p>
<div class="highlight"><pre><code class="language-r" data-lang="r">buildindex<span class="p">(</span>basename<span class="o">=</span><span class="s">&quot;chr1&quot;</span><span class="p">,</span> reference<span class="o">=</span><span class="s">&quot;chr1.fa&quot;</span><span class="p">)</span>
</code></pre></div>
<h3 id="step-2-align-reads-to-reference-genome">Step 2: Align reads to reference genome</h3>

<p>Now, we align the RNA-Seq reads in the FASTQ files to the reference genome:</p>
<div class="highlight"><pre><code class="language-r" data-lang="r">align<span class="p">(</span>index<span class="o">=</span><span class="s">&quot;chr1&quot;</span><span class="p">,</span> readfile1<span class="o">=</span><span class="kp">list.files</span><span class="p">(</span>pattern<span class="o">=</span><span class="s">&quot;.fastq.gz$&quot;</span><span class="p">))</span>
</code></pre></div>
<h3 id="step-3-count-the-number-of-reads-aligned-to-each-gene">Step 3: Count the number of reads aligned to each gene</h3>

<p>The next step is to count the number of reads that overlap genomic features of interest (eg. genes).</p>

<p>Here, we are using the inbuilt hg19 genes annotation. You have the option of providing other annotations in formats such as <a href="https://en.wikipedia.org/wiki/Gene_transfer_format">GTF</a>.</p>

<p>This step generates a table of read counts, with genes as rows and samples as colums:</p>
<div class="highlight"><pre><code class="language-r" data-lang="r">fCounts <span class="o">&lt;-</span> featureCounts<span class="p">(</span>files<span class="o">=</span><span class="kp">list.files</span><span class="p">(</span>pattern<span class="o">=</span><span class="s">&quot;.BAM$&quot;</span><span class="p">),</span> annot.inbuilt<span class="o">=</span><span class="s">&quot;hg19&quot;</span><span class="p">)</span>
</code></pre></div>
<p>Remove the file extensions from the sample names (Optional)</p>
<div class="highlight"><pre><code class="language-r" data-lang="r"><span class="kp">colnames</span><span class="p">(</span>fCounts<span class="o">$</span>counts<span class="p">)</span> <span class="o">&lt;-</span> <span class="kp">substr</span><span class="p">(</span><span class="kp">colnames</span><span class="p">(</span>fCounts<span class="o">$</span>counts<span class="p">),</span> start<span class="o">=</span><span class="m">1</span><span class="p">,</span> stop<span class="o">=</span><span class="m">2</span><span class="p">)</span>
</code></pre></div><div class="highlight"><pre><code class="language-r" data-lang="r"><span class="kp">head</span><span class="p">(</span>fCounts<span class="o">$</span>counts<span class="p">)</span>
</code></pre></div><div class="highlight"><pre><code class="language-text" data-lang="text">##           a1 a2 b1 b2
## 653635    78 79 65 61
## 100422834  0  0  0  0
## 645520     0  0  0  0
## 79501      0  0  0  0
## 729737    11  8  4  3
## 100507658  2  1  0  1
</code></pre></div>
<h3 id="step-4-filter-and-plot-the-data">Step 4: Filter and plot the data</h3>

<p>We want to remove the genes not expressed, or expressed at low-levels before normalising and testing for differential expression.</p>

<p>First, we will create a <code>DGEList</code> object using edgeR  </p>
<div class="highlight"><pre><code class="language-r" data-lang="r">dge <span class="o">&lt;-</span> DGEList<span class="p">(</span>counts<span class="o">=</span>fCounts<span class="o">$</span>counts<span class="p">)</span>
</code></pre></div>
<p>To remove genes expressed at relatively low-levels, we (arbitrarily) limit the subset the data to include only the genes to which at least 1 read per million reads mapped (1 cpm) in at least 2 samples.</p>
<div class="highlight"><pre><code class="language-r" data-lang="r">thresholdCPM <span class="o">&lt;-</span> <span class="kp">rowSums</span><span class="p">(</span>cpm<span class="p">(</span>dge<span class="p">)</span> <span class="o">&gt;</span> <span class="m">1</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="m">2</span>
dge <span class="o">&lt;-</span> dge<span class="p">[</span>thresholdCPM<span class="p">,</span> <span class="p">]</span>
</code></pre></div>
<p>It is also a very good idea plot the sample relations using a multi-dimensional scaling plot. Here, the <code>plotMDS</code> function shows the <em>distance</em> between samples in terms of the log2 fold-change. As expected, we can see that the distance between the &ldquo;a&rdquo; and &ldquo;b&rdquo; group samples is much greater than distance between the samples within each group (Note the different axis scales). For more details, type <code>?plotMDS</code></p>
<div class="highlight"><pre><code class="language-r" data-lang="r">plotMDS<span class="p">(</span>dge<span class="p">)</span>
</code></pre></div>
<p><img src="/images/quickRNAseq/unnamed-chunk-14-1.png" alt=""> </p>

<h3 id="step-5-transform-and-normalise-the-data">Step 5: Transform and normalise the data</h3>

<p>Now we can transform and normalise the data so we can test for differences between our groups. To do this we can use functions from the edgeR and Limma packages.</p>

<p>Here the <code>voom</code> function in the Limma package is used to transform the data. The resulting data set will have similar properties to normalised microarray data allowing the use of linear modelling.</p>

<p>You can also inspect at a couple of plots, one of the mean-variance trend generated by the <code>voom</code> function</p>
<div class="highlight"><pre><code class="language-r" data-lang="r">voomDat <span class="o">&lt;-</span> voom<span class="p">(</span>dge<span class="p">,</span> plot<span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
</code></pre></div>
<p><img src="/images/quickRNAseq/unnamed-chunk-15-1.png" alt=""> </p>

<p>and another MDS plot as above, after the voom normalisation-transformation step.</p>
<div class="highlight"><pre><code class="language-r" data-lang="r">plotMDS<span class="p">(</span>voomDat<span class="p">)</span>
</code></pre></div>
<p><img src="/images/quickRNAseq/unnamed-chunk-16-1.png" alt=""> </p>

<h3 id="step-6-test-for-differential-expression">Step 6: Test for differential expression</h3>

<p>First, we set up the design matrix. The &ldquo;A&rdquo; and &ldquo;B&rdquo; letters below correspond to the group for each samples.</p>
<div class="highlight"><pre><code class="language-r" data-lang="r">design <span class="o">&lt;-</span> model.matrix<span class="p">(</span><span class="o">~</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;A&quot;</span><span class="p">,</span> <span class="s">&quot;A&quot;</span><span class="p">,</span> <span class="s">&quot;B&quot;</span><span class="p">,</span> <span class="s">&quot;B&quot;</span><span class="p">))</span>
</code></pre></div>
<p>Now we can fit the linear models and test for differential expression using the <code>eBayes</code> and <code>lmFit</code> functions in Limma.</p>
<div class="highlight"><pre><code class="language-r" data-lang="r">fit <span class="o">&lt;-</span> eBayes<span class="p">(</span>lmFit<span class="p">(</span>voomDat<span class="p">,</span> design<span class="p">))</span>
</code></pre></div>
<p>Look at the top 10 differentially expressed genes:</p>
<div class="highlight"><pre><code class="language-r" data-lang="r">topTable<span class="p">(</span>fit<span class="p">,</span> coef<span class="o">=</span><span class="m">2</span><span class="p">,</span> n<span class="o">=</span><span class="m">10</span><span class="p">)</span>
</code></pre></div><div class="highlight"><pre><code class="language-text" data-lang="text">##            logFC   AveExpr         t      P.Value   adj.P.Val        B
## 6900    9.455124  7.890513  66.60225 1.834301e-06 0.001821007 4.941600
## 23208   5.616564 10.139017  46.46938 6.301567e-06 0.001821007 4.854583
## 26052   8.492326  7.409150  55.38321 3.452940e-06 0.001821007 4.493912
## 336    -8.173743  7.415885 -53.10933 3.986706e-06 0.001821007 4.451447
## 1063   -6.224667  8.026424 -42.71201 8.412690e-06 0.001821007 4.396636
## 23114   4.904982  9.419246  37.49921 1.313985e-05 0.001946317 4.348103
## 2563    8.259447  7.292882  45.92993 6.558870e-06 0.001821007 4.205470
## 388662  4.893855  9.189999  34.90678 1.679327e-05 0.002108754 4.149100
## 79727  -7.665938  7.162096 -45.44793 6.800354e-06 0.001821007 4.120870
## 11004  -5.613298  7.720785 -35.91942 1.522678e-05 0.002059129 3.962893
</code></pre></div>
<h3 id="summary">Summary</h3>

<p>It <em>can</em> be relatively easy to move from raw data (FASTQ files) to differential expression statistics <em>at the gene level</em>. However, in this example, the default options have been used for most functions, and these might not be appropriate for your data set. All of the R packages used above have extensive and well-written documentation, and with some reading, you should be able to achieve much more sophisticated analyses.</p>

<h3 id="notes-and-references">Notes and references</h3>

<p>This example was adapted from, and is similar to the case study that can be found at http://bioinf.wehi.edu.au/RNAseqCaseStudy/</p>

<p><strong>Rsubread</strong> <a href="http://www.bioconductor.org/packages/release/bioc/html/Rsubread.html">www.bioconductor.org/packages/release/bioc/html/Rsubread.html</a></p>

<p><strong>edgeR</strong> <a href="http://www.bioconductor.org/packages/release/bioc/html/edgeR.html">www.bioconductor.org/packages/release/bioc/html/edgeR.html</a></p>

<p><strong>limma</strong> <a href="http://www.bioconductor.org/packages/release/bioc/html/limma.html">www.bioconductor.org/packages/release/bioc/html/limma.html</a></p>

<p>The analysis methods used here were also used in the <a href="http://www.nature.com/nbt/journal/v32/n9/full/nbt.2957.html">comprehensive assessment of RNA-seq accuracy, reproducibility and information content by the Sequencing Quality Control Consortium</a></p>

<h3 id="r-session-info">R session info</h3>

<p>The platform and package versions I used to develop this example. </p>
<div class="highlight"><pre><code class="language-r" data-lang="r">sessionInfo<span class="p">()</span>
</code></pre></div><div class="highlight"><pre><code class="language-text" data-lang="text">## R version 3.1.1 (2014-07-10)
## Platform: x86_64-apple-darwin13.1.0 (64-bit)
## 
## locale:
## [1] en_AU.UTF-8/en_AU.UTF-8/en_AU.UTF-8/C/en_AU.UTF-8/en_AU.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] edgeR_3.8.4     limma_3.22.1    Rsubread_1.16.1 knitr_1.8      
## 
## loaded via a namespace (and not attached):
## [1] codetools_0.2-9 digest_0.6.4    evaluate_0.5.5  formatR_1.0    
## [5] stringr_0.6.2   tools_3.1.1
</code></pre></div>
    
    <!--<a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-sa/4.0/88x31.png" /></a><br />This work by <span xmlns:cc="http://creativecommons.org/ns#" property="cc:attributionName">Sam Buckberry</span> is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>.-->
</article>

    </main>
    
</body>

</html>
