<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" href="/css/main.css">
    <title>Sam Buckberry</title>
</head>
<body>
    <main>
        <header class="site-header">
            <h1 class="site-title"><a href="/">Sam Buckberry</a></h1>
            <nav class="site-nav">
                <a href="/about/" class="">About</a>
                <a href="/publications/" class="">Publications</a>
                <a href="/cv/" class="">CV</a>
                <!--
                <a href="/recipes/" class="">[R]ecipes</a>
                -->

            </nav>
        </header>
        <article class="post">
    <h2 class="post-header">Wrappers for running a co-expression network analysis (WGCNA)</h2>
    <p class="post-meta">2015-07-23</p>
    <h1 id="run-wgcna">RUN-WGCNA</h1>
<p>### Wrapper functions to perform a quick preliminary weighted-gene co-expression network analysis</p>

<h3 id="background">Background</h3>
<p>Weighted correlation network analysis, also known as weighted gene co-expression network analysis (WGCNA), is a widely used data mining method especially for studying biological networks based on pairwise correlations between variables. While it can be applied to most high-dimensional data sets, it has been most widely used in genomic applications. The WGCNA method is extremely powerful for analysing gene expression data for deriving results and insights that are not possible with typical <em>differential expression</em> experiments.</p>

<p>The WGCNA methods and algorithms can be implemented using the the WGCNA R package <a href="https://cran.r-project.org/web/packages/WGCNA/index.html">available on CRAN</a> and there are <a href="http://labs.genetics.ucla.edu/horvath/CoexpressionNetwork/Rpackages/WGCNA/Tutorials/index.html">extensive tutorials and examples</a> available.</p>

<p>Although these tutorials are excellent, the learning curve is steep, and they require the analyst to be reasonably proficient with R programming. As I have been regularly asked how to perform WGCNA, I thought I would share some wrapper scripts of some of the main functions that help get started with WGCNA.</p>

<p>I must emphasise that using these functions will not give the analyst much insight into how this method works. Nor will they allow you to fine-tune your analysis. If the user requires more functionality or in-depth analysis, I strongly urge they follow the WGCNA tutorials and manuals.</p>

<p>For a more in-depth description on WGCNA, see the <a href="https://en.wikipedia.org/wiki/Weighted_correlation_network_analysis">WGCNA wikipedia page</a> and the <a href="[http://labs.genetics.ucla.edu/horvath/CoexpressionNetwork/Rpackages/WGCNA/">WGCNA home page</a>.</p>

<h2 id="set-up">Set up</h2>
<p>First you need to ensure you have all the required R libraries to run these wrapper functions that can be found at <a href="https://github.com/SamBuckberry/RUN-WGCNA">RUN-WGCNA GitHub Repo</a>.</p>

<p><code class="highlighter-rouge">r
source("http://bioconductor.org/biocLite.R")
biocLite(c("AnnotationDbi", "impute", "GO.db", "preprocessCore"))
install.packages(pkgs=c("WGCNA", "flashClust", "ggplot2"))
</code></p>

<p>Load the libraries and wrapper scripts</p>

<p><code class="highlighter-rouge">r
source("https://raw.githubusercontent.com/SamBuckberry/RUN-WGCNA/master/WGCNA_functions.R")
library(reshape2)
library(ggplot2)
</code></p>

<p>Load the example data</p>

<p><code class="highlighter-rouge">r
dat &lt;- read.table(file=unz("exprsExtimates.csv.zip",filename="exprsExtimates.csv"), header=TRUE, row.names=1, sep=",", )
</code>
This example dataset is normalised microarray data from a human cell line experient designed to look at the gene expression differences in cells in different oxygen concentrations on different growth media.</p>

<h2 id="running-a-weighted-co-expression-analysis">Running a weighted co-expression analysis</h2>

<h3 id="step-1-determine-levels-of-connectivity-at-different-powers">Step 1. Determine levels of connectivity at different powers.</h3>
<p>Constructing  a  weighted  gene  network  requires the  choice  of  the  soft thresholding  power to  which  co-expression similarity is raised to calculate adjacency. The <code class="highlighter-rouge">determineSoftPowerWGCNA</code> function returns a set of network indices that should be inspected, for example:</p>

<p><code class="highlighter-rouge">r
determineSoftPowerWGCNA(data1=dat, outFile="plots/powerPlots.png", propGenes=.1)
</code>
<img src="https://github.com/SamBuckberry/RUN-WGCNA/blob/master/plots/powerPlots.png?raw=true" alt="Drawing" style="width: 600px;" /></p>

<p>These two plots show the summary network indices (y-axes) as functions of the soft-thresholding power (x-axes). Numbers in the plots indicate the cor- responding soft-thresholding powers. Plots indicate that approximate scale-free topology is attained at a power of 6. Therefore, the we will <em>weight</em> the network with a power of 6.</p>

<p>The <code class="highlighter-rouge">propGenes</code> argument specifies the proportion of genes to be included in the analysis. If this number is &gt; 1, ir selects the proportion of genes with the highest variance across samples. For example, <code class="highlighter-rouge">propGenes=0.33</code> will select the top 33% of most variable genes. This is desirable if the memory of your computer cannot handle larger datasets. However, the memory requirements can also be dealt with by using the <code class="highlighter-rouge">blockwiseModules</code> function in the WGCNA package. If your dataset has been through the appropriate filtering steps and you have the computational power, select <code class="highlighter-rouge">propGenes=1</code>.</p>

<h3 id="step-2-assemble-the-nework-with-the-chosen-power">Step 2. Assemble the nework with the chosen power.</h3>

<p><code class="highlighter-rouge">r
net &lt;- runWGCNA(data1=dat, propGenes=.1, softPower=3, signedNetwork=TRUE)
rm(dat)
</code>
The  functions in this wrapper have many  parameters,  and  in  this  example  most  of  them  are  left  at  their  default  value.</p>

<h3 id="step-3-look-at-the-modules-with-different-dynamic-tree-cut-heights">Step 3. Look at the modules with different dynamic tree cut heights.</h3>

<p><code class="highlighter-rouge">r
plotModulesCut(referenceDataset=net, outFile="plots/modClusterPlot.pdf")
</code>
<img src="https://github.com/SamBuckberry/RUN-WGCNA/blob/master/plots/modClusterPlot.png?raw=true" alt="Drawing" style="width: 900px;" /></p>

<p>Average linkage hierarchical clustering dendrogram of genes based on gene expression topological overlap. Modules of co-expressed genes were assigned colours which are represented in the horizontal bars below the dendrogram. The different rows of bars represent different dynamic tree cut heights.</p>

<h3 id="step-4-calculate-the-the-module-eigengenes">Step 4. Calculate the the module eigengenes</h3>

<p><code class="highlighter-rouge">r
e1 &lt;- calculateModuleEigengenes(referenceDataset=net, split=2)
</code></p>

<p>This step summarises each module by calculating the module eigengene for each sample, which is the first principal component of gene expression values for the module. Therefore, the eigengene represents a weighted average of gene ex- pression.</p>

<p>If I choose to use split=1 (the top row in the above dendrogram figure) there will be a small number of large modules in the network. Conversely, if I choose split=4, there will be a larger number of small co-expression modules. Depending on the purpose of the analysis, it is sometimes preferable to choose a larger number of small modules—in such a case, split values of 3-4 would be chosen.</p>

<h3 id="step-5-exploring-the-results-by-plotting-the-eigengene-expression-for-each-experimental-group-and-co-expression-module">Step 5. Exploring the results by plotting the eigengene expression for each experimental group and co-expression module.</h3>

<p>The example data used here were generated from a human cell line experiement designed to examine the effects of oxygen concentration and growth media on gene expression. There are 3 different oxygen concentrations (1%, 5%, 20%) and two different growth media (Plastic and Matrigel). The sample names represent these variables and the experimental replicate. Therefore, we can extract some sample information from the sample names and generate some meaningful plots.</p>

<p><code class="highlighter-rouge">r
groups &lt;- gsub('.{2}$', '', x=rownames(e1))
groups &lt;- t(data.frame(strsplit(groups, split=".", fixed=TRUE)))
colnames(groups) &lt;- c("Oxygen", "Media")
groups[ ,1] &lt;- gsub('X', '', groups[ ,1])
rownames(groups) &lt;- NULL
e1 &lt;- cbind(e1, groups)
</code></p>

<p>Then reshape the data for easy plotting with ggplot2.</p>

<p><code class="highlighter-rouge">r
melted &lt;- melt(data=e1, id.vars=c("Oxygen", "Media"))
melted$Oxygen &lt;- gsub(pattern="pc", replacement="", melted$Oxygen)
melted &lt;- melted[order(as.numeric(melted$Oxygen)), ]
melted$variable &lt;- gsub("ME", "", melted$variable)
</code></p>

<p>…and then plot the data.</p>

<p><code class="highlighter-rouge">r
png(file="plots/eigengenBoxplot.png")
q &lt;- qplot(data=melted, y=value, x=Oxygen, facets=Media~variable, 
           geom=c("boxplot", "point"), ylab="Eignegene expression",
           colour=Oxygen)
q + scale_x_discrete(limits=c("1","5","20"))
dev.off()
</code>
<img src="https://github.com/SamBuckberry/RUN-WGCNA/blob/master/plots/eigengenBoxplot.png?raw=true" alt="Drawing" /></p>

<p>These boxplots show the eigenene ‘expression’ (y-axis) for each module (defined by colors) in panel columns for varying oxygen concentrations (x-axis) on different growth media (panel rows). We can see for the plastic media, the ‘brown’ module shows notable eigengene expression differences at 1% oxygen which are much more profound than the differences observed for the same oxygen concentration on the matrigel media. This suggests there are differences in the media that induce differences in co-expression in 1% oxygen.</p>

<p>Get the top module hub genes…</p>

<p><code class="highlighter-rouge">r
hubs &lt;- moduleHubGenes(referenceDataset=net, MEs=calculateModuleEigengenes(referenceDataset=net, split=2), nGenes=10, split=2)
</code></p>

<p>…and list the top array genes for the brown module</p>

<p><code class="highlighter-rouge">r
hubs[ ,colnames(hubs) == "brown"]
</code></p>

<p><code class="highlighter-rouge">
##  [1] "ILMN_1736585" "ILMN_1757384" "ILMN_2383934" "ILMN_2279635"
##  [5] "ILMN_1783684" "ILMN_1669584" "ILMN_1714820" "ILMN_1793384"
##  [9] "ILMN_2189993" "ILMN_1777261"
</code></p>

    
    <!--<a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-sa/4.0/88x31.png" /></a><br />This work by <span xmlns:cc="http://creativecommons.org/ns#" property="cc:attributionName">Sam Buckberry</span> is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>.-->
</article>

    </main>
    
</body>

</html>
